pkg/controller/controller_instance.go

```diff
@@ -1137,20 +1157,22 @@ func (c *controller) updateServiceInstanceCondition(
 	return err
 }
 
-// prepareServiceInstanceStatus sets the instance's observed generation
-// and clears the conditions, preparing it for any status updates that can occur
-// during the further processing.
-// It doesn't send the update request to server.
-func (c *controller) prepareServiceInstanceStatus(toUpdate *v1beta1.ServiceInstance) {
+// recordServiceInstanceObservedGeneration updates the instance's observed generation
+// and clears the conditions.
+func (c *controller) recordServiceInstanceObservedGeneration(toUpdate *v1beta1.ServiceInstance) (*v1beta1.ServiceInstance, error) {
 	toUpdate.Status.ObservedGeneration = toUpdate.Generation
-	resetServiceInstanceCondition(
+	setServiceInstanceCondition(
 		toUpdate,
 		v1beta1.ServiceInstanceConditionReady,
+		v1beta1.ConditionFalse,
+		startProcessingReason,
+		startProcessingMessage,
 	)
 	resetServiceInstanceCondition(
 		toUpdate,
 		v1beta1.ServiceInstanceConditionFailed,
 	)
+	return c.updateServiceInstanceStatus(toUpdate)
```

Comment:
Initially I didn't want to sent an update request there (and instead send it with the first status update), but:
1. The `updateServiceInstanceReferences()` updates only references but re-reads the entire instance from the server, overwriting the ObservedGeneration.
Doesn't it lead to re-reading the spec too, thus potentially reading the newer spec than the recorded ObservedGeneration? @kibbles-n-bytes @staebler see [controller_instance.go#L1030](https://github.com/kubernetes-incubator/service-catalog/blob/ca83d186a762b944ed6b33b7493027e241b8d9fa/pkg/controller/controller_instance.go#L1030)
If it does, then we can't set `ObservedGeneration` separately from storing `InProgressProperties`, i.e. ObservedGeneration should be only updated in `recordStartOfServiceInstanceOperation()` and not earlier...
2. if the status update will fail later, we won't update ObservedGeneration until the next iteration, thus the status will be obsolete for a while

Reply:
Actually still not possible, because:
```go
rh.inProgressProperties = &v1beta1.ServiceInstancePropertiesState{
  ClusterServicePlanExternalName: servicePlan.Spec.ExternalName,
  ClusterServicePlanExternalID:   servicePlan.Spec.ExternalID,
  ...
```
so annoying...

Reply:
@nilebox 

> re-reads the entire instance from the server, overwriting the ObservedGeneration

What do you mean be overwriting the `ObservedGeneration`? If that field has been changed in the instance held by the controller, then it will be written to the server. My reading of what you wrote is that you think that the change to the `ObservedGeneration` made by the controller will be overwritten by whatever is currently held in storage prior to the update. So, for example, if the `ObservedGeneration` in storage is 2, the controller set it to 3, then `updateServiceInstanceReferences` will end up setting it back to 2. My understanding is different. I would expect `updateServiceInstanceReferences` to cause the `ObservedGeneration` in storage to change to 3. The `ObservedGeneration` of the instance held by the controller will remain 3.

Reply:
> Doesn't it lead to re-reading the spec too, thus potentially reading the newer spec than the recorded ObservedGeneration?

@nilebox It will **not** lead to reading in a newer spec. If the spec is newer in storage, then the update will fail with a conflict. The reconciliation will end there with no changes affected. The instance will be re-queued, at which time the controller will attempt to reconcile the newer spec.

Reply:
@staebler thanks, I completely forgot that update will fail with conflict, so there is no possibility of re-reading a newer spec.

>  I would expect updateServiceInstanceReferences to cause the ObservedGeneration in storage to change to 3. 

No, this method is a subresource IIRC, and it will ignore the status (won't update ObservedGeneration) and will overwrite any changes to the status when instance is re-read. That's why we need to send a status update with updated ObservedGeneration separately before updating references. Please correct me if I'm missing something again.

Reply:
Alternatively, we can just remember the local changes to status and overwrite it again on the controller side. Then we don't have to make a separate call for updating ObservedGeneration - it will be updated with the first status update (i.e. in case of any error or when we record the operation)